core/src/main/scala/org/apache/spark/deploy/worker/WorkerWatcher.scala

```diff
@@ -45,7 +50,14 @@ private[spark] class WorkerWatcher(
   private val expectedAddress = RpcAddress.fromURIString(workerUrl)
   private def isWorker(address: RpcAddress) = expectedAddress == address
 
-  private def exitNonZero() = if (isTesting) isShutDown = true else System.exit(-1)
+  private def exitNonZero() =
+    if (isTesting) {
+      isShutDown = true
+    } else {
+      ThreadUtils.awaitResult(Future {
```

Comment:
This is more like a temporary workaround: if we hit the dead lock, we exist after 5 seconds. Is there a way to fully eliminate the dead lock so that we never hit it here?

cc @rednaxelafx @jiangxb1987 @Ngone51

Reply:
Maybe, using a global flag between `WorkerWatcher` and `CoarseGrainedExecutorBackend` to indicate whether `System.exit` has been called?

Reply:
Discussed with @Ngone51 offline, we can set a flag in `CoarseGrainedExecutorBackend`, and check that flag in `WorkerWatcher`.

Reply:
Great! please open a PR and we can start review.

Reply:
+1 for fixing this specific issue better.
Note that in general case, these sorts of interactions while in shutdown hooks, are painful: we would need to be careful with all `exit` calls.

Reply:
@cloud-fan @Ngone51 Sorry, I did not find a good solution for this issue.

Reply:
> Probably add something in Utils for exit - which checks `inShutdown` and avoids making System.exit call entirely ? (and require all exit calls to go through Utils)

That would mitigate the issue. But seems like `System.exit` and shutdownhook are not executed atomically, so the race condition would still exist, right?

Reply:
A simplified code:
```
    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
      System.exit(-1);
    }));
    System.exit(-2);
```

And I prefer to add `inShutdown` flag in `SignalUtils.ActionHandler` as it's a jvm flag.

Reply:
I don't understand the code. Would it result in an endless loop since `System.exit(-1)` invokes shutdownhook too?

Reply:
BTW, I have submitted the new PR(https://github.com/apache/spark/pull/33028). Please help review. And we can have more discussion there.